pkg/sharings/sharings.go

```diff
@@ -288,104 +287,171 @@ func ShareDoc(instance *instance.Instance, sharing *Sharing, recStatus *Recipien
-					if err != nil {
-						return err
-					}
-					for _, row := range res.Rows {
-						values = append(values, row.ID)
-					}
+			// Selector-based sharing
+			values, err = sharingBySelector(instance, rule)
+			if err != nil {
+				return err
+			}
+		} else {
+			// Value-based sharing
+			values, err = sharingByValues(instance, rule)
+			if err != nil {
+				return err
+			}
+		}
+		err = sendData(instance, sharing, recStatus, values, rule)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
 
-				}
-			} else {
+// sharingBySelector returns the ids to share based on the Rule selector
+func sharingBySelector(instance *instance.Instance, rule permissions.Rule) ([]string, error) {
+	var values []string
+
+	// Particular case for referenced_by: use the existing view
+	if rule.Selector == consts.SelectorReferencedBy {
+		for _, val := range rule.Values {
+			// A referenced_by selector implies Values in the form
+			// ["refDocType/refId"]
+			parts := strings.Split(val, permissions.RefSep)
+			if len(parts) != 2 {
+				return nil, ErrBadPermission
+			}
+			refType := parts[0]
+			refID := parts[1]
+			req := &couchdb.ViewRequest{
+				Key:    []string{refType, refID},
+				Reduce: false,
+			}
+			var res couchdb.ViewResponse
+			err := couchdb.ExecView(instance,
+				consts.FilesReferencedByView, req, &res)
+			if err != nil {
+				return nil, err
+			}
+			for _, row := range res.Rows {
+				values = append(values, row.ID)
+			}
 
-				// Create index based on selector to retrieve documents to share
-				indexName := "by-" + rule.Selector
-				index := mango.IndexOnFields(docType, indexName,
-					[]string{rule.Selector})
-				err := couchdb.DefineIndex(instance, index)
+		}
+	} else {
+		// Create index based on selector to retrieve documents to share
+		indexName := "by-" + rule.Selector
+		index := mango.IndexOnFields(rule.Type, indexName,
+			[]string{rule.Selector})
+		err := couchdb.DefineIndex(instance, index)
+		if err != nil {
+			return nil, err
+		}
+
+		var docs []couchdb.JSONDoc
+
+		// Request the index for all values
+		// NOTE: this is not efficient in case of many Values
+		// We might consider a map-reduce approach in case of bottleneck
+		for _, val := range rule.Values {
+			err = couchdb.FindDocs(instance, rule.Type,
+				&couchdb.FindRequest{
+					UseIndex: indexName,
+					Selector: mango.Equal(rule.Selector, val),
+				}, &docs)
+			if err != nil {
+				return nil, err
+			}
+			// Save returned doc ids
+			for _, d := range docs {
+				values = append(values, d.ID())
+			}
+		}
+	}
+	return values, nil
+}
+
+// sharingByValues returns the ids to share based on the Rule values
+func sharingByValues(instance *instance.Instance, rule permissions.Rule) ([]string, error) {
+	var values []string
+
+	// Iterate on values to detect directory sharing
+	for _, val := range rule.Values {
+		if rule.Type == consts.Files {
+			fs := instance.VFS()
+			dirDoc, _, err := fs.DirOrFileByID(val)
```

Comment:
However, at this point we truly don't know if it's a directory or a file; so it might be nicer to use `DirOrFileByID` to avoid dealing with an error from `DirByID` in case it's a file?